### PR TITLE
refactor: use webpack instead of webpack-sources

### PR DESCRIPTION
**Pull request by yoannmoinet**
Coming from the discussion https://github.com/unjs/community/discussions/37

Having a dependency on `webpack-sources` forces plugin developers to break the compatibility `webpack`/`webpack-sources` by having them declare a dependency on `webpack-sources` instead of a peerDependency on `webpack`.

Instead, we could require `webpack-sources` **as** `webpack` using:

```javascript
import { createRequire } from 'module'
[...]
const webpackRequire = createRequire(require.resolve('webpack'))
const RawSource = webpackRequire('webpack-sources').RawSource
```

And remove any dependency on `webpack-sources` or `webpack`.

The change "should" be mostly invisible.

- [x] Confirmed in both webpack 4 and 5 projects.

### CODE REVIEW

**Comment by yoannmoinet** [comment]
I will have a look at the failing test later today.

File: package.json [hunk 1]
```diff
@@ -46,10 +46,10 @@
     "test:build": "jiti scripts/buildFixtures.ts"
   },
   "peerDependencies": {
-    "webpack-sources": "^3"
+    "webpack": ">= 4.x <= 5.x"
```
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply package.json:49]: No strong feelings.
I'll update, it is a bit more readable indeed.

**Comment by yoannmoinet** [comment]
🤔 
I'm having difficulties to understand why the test started failing here after the `main` merge.
I can't reproduce it locally.
Anyone can assist here?

**Comment by yoannmoinet** [comment]
So sorry for hammering, but I'm stuck on this on a separate PR.
Any update @sxzz? 
Also summoning @antfu, maybe you'll have an idea of what the issue may be.

**Comment by sxzz** [comment]
It's OK. I'll check it out when I have time.

**Comment by sxzz** [comment]
I think the PR is incompatible with Webpack 4.

image



image

**Comment by yoannmoinet** [comment]
Thank you so much for looking into it.

I'll have a better look, I don't understand why it's not _always_ failing though.
I thought the fallback here was specifically for this case where webpack doesn't use `webpack-sources` (aka in webpack 4).

https://github.com/unjs/unplugin/blob/dd9d184dc69bcf97708a0b2dc1e33f58b8dd14e1/src/webpack/context.ts#L48-L58

I'll double check how webpack 4 works there, and if it actually needs `webpack-sources` or not.

**Comment by yoannmoinet** [comment]
So, I couldn't find a good solution for webpack 4.

What I propose, to keep the best compatibility possible between `webpack` and `webpack-sources` is:

- Add an optional peerDependency on `webpack-sources@^1.4.1` ([version used by latest webpack@4](https://github.com/webpack/webpack/blob/dfffd6a241bf1d593b3fd31b4b279f96f4a4aab1/package.json#L30)).
- If `require('webpack').sources` exists, use it.
- If `require('webpack').sources` doesn't exist, use `require('webpack-sources').RawSource` instead.
- Add documentation about adding a dependency on `webpack-sources:^1.4.1` for plugin developers so they don't break dependencies or side-load a potentially non-existing dependency.

This way we keep the compatibility most of the time, and do our best for webpack 4.

Cf: https://github.com/webpack/webpack/pull/11345

> Plugins should consume require("webpack").sources instead of require("webpack-sources") to avoid version conflicts for persistent caching

WDYT @sxzz ?

**Comment by yoannmoinet** [comment]
Ok, so I went a different route after all.
It appears that you can actually require a package as a different dependent.

In our case we wanted to require the same `webpack-sources` as our peer `webpack`.

Doing: 
```javascript
import { createRequire } from 'module'
[...]
const webpackRequire = createRequire(require.resolve('webpack'))
const RawSource = webpackRequire('webpack-sources').RawSource
```

Makes it require the correct `webpack-sources` that our peer `webpack` expects.
And, we don't need to declare any of them as `peerDependencies` anymore.

(But, plugin developers should still declare their plugin's bundler as `peerDependencies`)

And it seems to pass the tests now 🙌

**Comment by sxzz** [comment]
I need to test the PR more, please hold on.

**Comment by yoannmoinet** [comment]
So sorry to ping you again @sxzz.
Did you have a chance to test the PR as you wanted? Can I help you in any way?

**Review by sxzz — APPROVED** [review]
LGTM

**Pull request merged**